src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala

```diff
@@ -209,26 +230,58 @@ class TezosPlatformDiscoveryOperations(
     } yield {
-      val invalidDataTypeValidationResult = if (!attrOpt.exists(attr => canQueryType(attr.dataType))) Some(InvalidAttributeDataType(column)) else None
-      val highCardinalityValidationResult = if (!isLowCardinality(attrOpt.flatMap(_.cardinality))) Some(HighCardinalityAttribute(column)) else None
-      val validationErrors = List(invalidDataTypeValidationResult, highCardinalityValidationResult).flatten
-
-      val res = Either.cond(
-        test = validationErrors.isEmpty,
-        right = attrOpt.map(attr => makeAttributesQuery(tableName, attr.name, withFilter)).toList.sequence.map(_.flatten),
-        left = Future.successful(validationErrors)
-      )
+      val res = (attributesCacheConfig, withFilter) match {
+        case (Some(AttributeCacheConfiguration(cached, minMatchLength, maxResultLength)), Some(attributeFilter)) if cached && attributeFilter.length >= minMatchLength  =>
+          Right(getAttributeValuesFromCache(tableName, column, attributeFilter, maxResultLength))
+
+        case (Some(AttributeCacheConfiguration(cached, minMatchLength, _)), Some(attributeFilter)) if cached && attributeFilter.length < minMatchLength =>
+          Left(Future.successful(List(InvalidAttributeFilterLength(column, minMatchLength))))
+
+        case _ =>
+          val invalidDataTypeValidationResult = if (!attrOpt.exists(attr => canQueryType(attr.dataType))) Some(InvalidAttributeDataType(column)) else None
+          val highCardinalityValidationResult = if (!isLowCardinality(attrOpt.flatMap(_.cardinality))) Some(HighCardinalityAttribute(column)) else None
+          val validationErrors = List(invalidDataTypeValidationResult, highCardinalityValidationResult).flatten
+          Either.cond(
+            test = validationErrors.isEmpty,
+            right = attrOpt.map(attr => makeAttributesQuery(tableName, attr.name, withFilter)).toList.sequence.map(_.flatten),
+            left = Future.successful(validationErrors)
+          )
+      }
+
       res.bisequence
     }
     result.flatten
   }
 
+  /** Gets attribute values from cache and updates them if necessary */
+  private def getAttributeValuesFromCache(tableName: String, columnName: String, attributeFilter: String, maxResultLength: Int): Future[List[String]] = {
+    attributeValuesCache.read.flatMap { cache =>
+      cache(makeKey(tableName, columnName)) match {
+        case (last, radixTree) if last + cacheTTL.toMillis > now =>
+          IO.pure(radixTree.filterPrefix(attributeFilter.toLowerCase).values.take(maxResultLength).toList)
+        case _ =>
+          for {
+            cache <- attributeValuesCache.take
+            attributeValues <- IO.fromFuture(IO(makeAttributesQuery(tableName, columnName, None)))
+            radixTree = RadixTree(attributeValues.map(x => x.toLowerCase -> x): _*)
+            cacheTmp = cache.updated(makeKey(tableName, columnName), (now, radixTree))
+            _ <- attributeValuesCache.put(cacheTmp)
```

Comment:
I'm not sure, but if more than one thread has concurrently `read` the cache and found it's stale, they might be doing a race to update the value, so multiple `take` might be serialized to update the value.

To avoid useless work, we could check again for the cache being invalid, right after the `take` call, and simpliy `put` back the value if another thread have updated it concurrently...

Reply:
really good catch!